atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/invariant/DurableWritesInvariantTest.java

```diff
@@ -0,0 +1,123 @@
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedDeleteTransactionAction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedTransaction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedWriteTransactionAction;
+import com.palantir.atlasdb.workload.util.AtlasDbUtils;
+import com.palantir.atlasdb.workload.workflow.ImmutableWorkflowHistory;
+import com.palantir.atlasdb.workload.workflow.WorkflowHistory;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class DurableWritesInvariantTest {
+
+    private static final String TABLE = "foo";
+    private static final TableReference TABLE_REFERENCE = TableReference.create(Namespace.DEFAULT_NAMESPACE, TABLE);
+
+    private static final WorkloadCell WORKLOAD_CELL_1 = ImmutableWorkloadCell.of(1, 2);
+    private static final WorkloadCell WORKLOAD_CELL_2 = ImmutableWorkloadCell.of(3, 4);
+    private static final Integer VALUE_ONE = 5;
+    private static final Integer VALUE_TWO = 3;
+
+    private TransactionManager manager;
+    private AtlasDbTransactionStore store;
+
+    @Before
+    public void before() {
+        manager = TransactionManagers.createInMemory(Set.of());
+        store = AtlasDbTransactionStore.create(
+                manager, Map.of(TABLE_REFERENCE, AtlasDbUtils.tableMetadata(ConflictHandler.SERIALIZABLE)));
+    }
+
+    @Test
+    public void undeletedCellsAreFoundAndDeletedAreIgnored() {
+        AtomicReference<Map<TableWorkloadCell, Integer>> undeletedCells = new AtomicReference<>();
+
+        DurableWritesInvariant durableWritesInvariant = new DurableWritesInvariant(_ignore -> {}, undeletedCells::set);
+        List<WitnessedTransaction> witnessedTransactions = List.of(
+                ImmutableWitnessedTransaction.builder()
+                        .addActions(WitnessedDeleteTransactionAction.of(TABLE, WORKLOAD_CELL_1))
+                        .startTimestamp(1)
+                        .commitTimestamp(2)
+                        .build(),
+                ImmutableWitnessedTransaction.builder()
+                        .addActions(WitnessedDeleteTransactionAction.of(TABLE, WORKLOAD_CELL_2))
+                        .startTimestamp(3)
+                        .commitTimestamp(4)
+                        .build());
+        WorkflowHistory history = ImmutableWorkflowHistory.builder()
+                .history(witnessedTransactions)
+                .transactionStore(store)
+                .build();
+
+        store.readWrite(List.of(WriteTransactionAction.of(TABLE, WORKLOAD_CELL_2, VALUE_ONE)));
+        durableWritesInvariant.accept(history);
+        assertThat(undeletedCells.get())
+                .containsExactlyEntriesOf(Map.of(TableWorkloadCell.of(TABLE, WORKLOAD_CELL_2), VALUE_ONE));
```

Comment:
nit/pedantry: technically we should be using `containsExactlyInAnyOrderEntriesOf` here, this version does assert on the iteration order. I realise we just have one cell, but someone copying this test might make mistakes